heron/packing/src/java/com/twitter/heron/packing/roundrobin/RoundRobinPacking.java

```diff
@@ -27,44 +26,263 @@
+ * value for com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_CPU_REQUESTED if exists, otherwise,
+ * (cpu for instances in container) + (cpu padding for heron internal process)
+ * <p>
+ * 5. The ram required for a container is calculated as:
+ * value for com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_RAM_REQUESTED if exists, otherwise,
+ * (ram for instances in container) + (ram padding for heron internal process)
+ * <p>
+ * 6. The ram required for one instance is calculated as:
+ * value in com.twitter.heron.api.Config.TOPOLOGY_COMPONENT_RAMMAP if exists, otherwise,
+ * - if com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_RAM_REQUESTED not exists:
+ * the default ram value for one instance
+ * - if com.twitter.heron.api.Config.TOPOLOGY_CONTAINER_RAM_REQUESTED exists:
+ * ((TOPOLOGY_CONTAINER_RAM_REQUESTED) - (ram padding for heron internal process)
+ * - (ram used by instances within TOPOLOGY_COMPONENT_RAMMAP config))) /
+ * (the # of instances in container not specified in TOPOLOGY_COMPONENT_RAMMAP config)
+ * 7. The pack() return null if PackingPlan fails to pass the safe check, for instance,
+ * the size of ram for an instance is less than the minimal required value.
  */
 public class RoundRobinPacking implements IPacking {
   private static final Logger LOG = Logger.getLogger(RoundRobinPacking.class.getName());
-  private static final long DEFAULT_DISK_PADDING = 12 * Constants.GB;
+
+  // TODO(mfu): Read these values from Config
+  public static final long DEFAULT_DISK_PADDING_PER_CONTAINER = 12L * Constants.GB;
+  public static final long DEFAULT_RAM_PADDING_PER_CONTAINER = 2L * Constants.GB;
+  public static final double DEFAULT_CPU_PADDING_PER_CONTAINER = 1;
+  public static final long MIN_RAM_PER_INSTANCE = 192L * Constants.MB;
+
+  // Use as a stub as default number value when getting config value
+  public static final String NOT_SPECIFIED_NUMBER_VALUE = "-1";
 
   protected TopologyAPI.Topology topology;
-  protected long stmgrRamDefault;
+
   protected long instanceRamDefault;
   protected double instanceCpuDefault;
   protected long instanceDiskDefault;
 
   @Override
   public void initialize(Config config, Config runtime) {
-    this.stmgrRamDefault = Context.stmgrRam(config);
+    this.topology = com.twitter.heron.spi.utils.Runtime.topology(runtime);
+
     this.instanceRamDefault = Context.instanceRam(config);
     this.instanceCpuDefault = Context.instanceCpu(config).doubleValue();
     this.instanceDiskDefault = Context.instanceDisk(config);
-    this.topology = Runtime.topology(runtime);
   }
 
   @Override
   public PackingPlan pack() {
-    Map<String, List<String>> containerInstancesMap = getBasePacking();
-    return fillInResource(containerInstancesMap);
+    // Get the instances' round-robin allocation
+    Map<String, List<String>> roundRobinAllocation = getRoundRobinAllocation();
+
+    // Get the ram map for every instance
+    Map<String, Map<String, Long>> instancesRamMap =
```

Comment:
Working with maps of Strings to Maps of Strings to Longs can get confusing and could lead to bugs. We should consider using more strongly typed objects, like a Container that has a Collection of Instances, which each has requiredRam. It would make the code much readable and have better compile time checks.